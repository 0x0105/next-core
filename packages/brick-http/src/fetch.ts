export type Interceptor = (
  req: Request,
  next: NextInterceptor,
  extra?: any
) => Promise<Response>;

export type NextInterceptor = (req: Request) => Promise<Response>;

export type RemoveInterceptor = () => void;

const interceptors: Interceptor[] = [];

export const fetch = (
  input: RequestInfo,
  init?: RequestInit,
  interceptorParams?: any
): Promise<Response> => {
  // Support older browsers which default credentials is "omit".
  // Ref https://github.com/whatwg/fetch/pull/585
  const req = new Request(
    input,
    Object.assign<RequestInit, RequestInit>(
      {
        credentials: "same-origin"
      },
      init
    )
  );
  let last = (req: Request): Promise<Response> => self.fetch(req);
  for (const inter of interceptors.slice().reverse()) {
    const next: NextInterceptor = last;
    last = req => inter(req, next, interceptorParams);
  }
  return last(req);
};

export const pushInterceptor = (
  interceptor: Interceptor
): RemoveInterceptor => {
  interceptors.push(interceptor);
  return () => {
    if (interceptors.includes(interceptor)) {
      interceptors.splice(interceptors.indexOf(interceptor), 1);
    }
  };
};
