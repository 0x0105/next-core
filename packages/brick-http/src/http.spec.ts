// @ts-ignore
import { fetch, __setReturnValue } from "./fetch";
import {
  http,
  HttpOptions,
  HttpFetchError,
  HttpParseError,
  HttpResponseError,
} from "./";

jest.mock("./fetch");

const spyOnFetch = fetch as jest.Mock;

type TestItem =
  | ["GET" | "DELETE" | "HEAD", string, [HttpOptions?]]
  | ["POST" | "PUT" | "PATCH", string, [any?, HttpOptions?]];

// Hack for snapshot testing.
(FormData.prototype as any).toJSON = function () {
  return "<FormData> " + JSON.stringify(Array.from(this.entries()));
};
(URLSearchParams.prototype as any).toJSON = function () {
  return "<URLSearchParams> " + this.toString();
};

describe("http", () => {
  afterEach(() => {
    spyOnFetch.mockClear();
  });

  const formData = new FormData();
  formData.append("for", "the throne");

  const batchTests: TestItem[] = [
    ["GET", "http://example.com", []],
    ["POST", "http://example.com", []],
    ["PUT", "http://example.com", []],
    ["PATCH", "http://example.com", []],
    ["DELETE", "http://example.com", []],
    ["HEAD", "http://example.com", []],
    [
      "GET",
      "http://example.com",
      [{ params: { lang: "zh-CN", type: null, category: undefined } }],
    ],
    [
      "GET",
      "http://example.com",
      [{ params: new URLSearchParams("lang=zh-CN") }],
    ],
    [
      "GET",
      "http://example.com",
      [
        {
          params: { lang: ["zh-CN", "en-US"], type: null, category: undefined },
        },
      ],
    ],
    [
      "GET",
      "http://example.com/for-good?token=secret",
      [{ params: { lang: "zh-CN", type: null, category: undefined } }],
    ],
    ["POST", "http://example.com", [{ for: "good" }]],
    [
      "POST",
      "http://example.com/for-good",
      [
        { for: "the throne" },
        { params: { lang: "zh-CN", type: null, category: undefined } },
      ],
    ],
    [
      "POST",
      "http://example.com/for-good",
      [
        { for: "the throne" },
        {
          params: new URLSearchParams("lang=zh-CN"),
        },
      ],
    ],
    [
      "POST",
      "http://example.com/for-good",
      [
        { for: "the throne" },
        { headers: { "x-requested-with": "XMLHttpRequest" } },
      ],
    ],
    [
      "POST",
      "http://example.com/for-good",
      [
        { for: "the throne" },
        { headers: new Headers({ "x-requested-with": "XMLHttpRequest" }) },
      ],
    ],
    ["POST", "http://example.com/for-good", ["for=the-throne"]],
    ["POST", "http://example.com/for-good", [formData]],
  ];

  it.each(batchTests)(
    "%s %s with %j should work",
    async (method: string, url: string, args: any[]) => {
      await http[method.toLowerCase() as "get"](url, ...args);
      expect(spyOnFetch.mock.calls[0]).toMatchSnapshot();
    }
  );

  it("should return raw text", async () => {
    __setReturnValue(Promise.resolve(new Response("raw-text")));
    const result = await http.get("http://example.com", {
      responseType: "text",
    });
    expect(result).toBe("raw-text");
  });

  it("should return blob", async () => {
    __setReturnValue(Promise.resolve(new Response(new Blob())));
    const result = await http.get("http://example.com", {
      responseType: "blob",
    });
    expect(result).toBeInstanceOf(Blob);
  });

  it("should return array buffer", async () => {
    __setReturnValue(Promise.resolve(new Response(new ArrayBuffer(1))));
    const result = await http.get("http://example.com", {
      responseType: "arrayBuffer",
    });
    expect(result).toBeInstanceOf(ArrayBuffer);
  });

  it("should throw a HttpFetchError", async () => {
    __setReturnValue(Promise.reject(new Error("mock error")));
    expect.assertions(1);
    try {
      await http.get("http://example.com");
    } catch (e) {
      expect(e).toBeInstanceOf(HttpFetchError);
    }
  });

  it("should throw a HttpResponseError", async () => {
    __setReturnValue(
      Promise.resolve(
        new Response('{"error":"oops"}', {
          status: 500,
        })
      )
    );
    expect.assertions(2);
    try {
      await http.get("http://example.com");
    } catch (e) {
      expect(e).toBeInstanceOf(HttpResponseError);
      expect(e.responseJson).toEqual({
        error: "oops",
      });
    }
  });

  it("should throw a HttpParseError", async () => {
    __setReturnValue(Promise.resolve(new Response("non-json")));
    expect.assertions(1);
    try {
      await http.get("http://example.com");
    } catch (e) {
      expect(e).toBeInstanceOf(HttpParseError);
    }
  });
});
