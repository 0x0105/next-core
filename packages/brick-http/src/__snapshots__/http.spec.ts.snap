// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`http DELETE http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "DELETE",
  },
  undefined,
]
`;

exports[`http GET http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "GET",
  },
  undefined,
]
`;

exports[`http GET http://example.com with [{"params":"<URLSearchParams> lang=zh-CN"}] should work 1`] = `
Array [
  "http://example.com/?lang=zh-CN",
  Object {
    "method": "GET",
  },
  undefined,
]
`;

exports[`http GET http://example.com with [{"params":{"lang":"zh-CN","type":null}}] should work 1`] = `
Array [
  "http://example.com/?lang=zh-CN&type=&category=",
  Object {
    "method": "GET",
  },
  undefined,
]
`;

exports[`http GET http://example.com with [{"params":{"lang":["zh-CN","en-US"],"type":null}}] should work 1`] = `
Array [
  "http://example.com/?lang=zh-CN&lang=en-US&type=&category=",
  Object {
    "method": "GET",
  },
  undefined,
]
`;

exports[`http GET http://example.com/for-good?token=secret with [{"params":{"lang":"zh-CN","type":null}}] should work 1`] = `
Array [
  "http://example.com/for-good?token=secret&lang=zh-CN&type=&category=",
  Object {
    "method": "GET",
  },
  undefined,
]
`;

exports[`http HEAD http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "HEAD",
  },
  undefined,
]
`;

exports[`http PATCH http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "PATCH",
  },
  undefined,
]
`;

exports[`http POST http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "POST",
  },
  undefined,
]
`;

exports[`http POST http://example.com with [{"for":"good"}] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "body": "{\\"for\\":\\"good\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
      },
    },
    "method": "POST",
  },
  undefined,
]
`;

exports[`http POST http://example.com/for-good with ["<FormData> [[\\"for\\",\\"the throne\\"]]"] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "<FormData> [[\\"for\\",\\"the throne\\"]]",
    "headers": Headers {
      "map": Object {},
    },
    "method": "POST",
  },
  undefined,
]
`;

exports[`http POST http://example.com/for-good with ["for=the-throne"] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "for=the-throne",
    "headers": Headers {
      "map": Object {
        "content-type": "application/x-www-form-urlencoded",
      },
    },
    "method": "POST",
  },
  undefined,
]
`;

exports[`http POST http://example.com/for-good with [{"for":"the throne"},{"headers":{"map":{"x-requested-with":"XMLHttpRequest"}}}] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
        "x-requested-with": "XMLHttpRequest",
      },
    },
    "method": "POST",
  },
  undefined,
]
`;

exports[`http POST http://example.com/for-good with [{"for":"the throne"},{"headers":{"x-requested-with":"XMLHttpRequest"}}] should work 1`] = `
Array [
  "http://example.com/for-good",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
        "x-requested-with": "XMLHttpRequest",
      },
    },
    "method": "POST",
  },
  undefined,
]
`;

exports[`http POST http://example.com/for-good with [{"for":"the throne"},{"params":"<URLSearchParams> lang=zh-CN"}] should work 1`] = `
Array [
  "http://example.com/for-good?lang=zh-CN",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
      },
    },
    "method": "POST",
  },
  undefined,
]
`;

exports[`http POST http://example.com/for-good with [{"for":"the throne"},{"params":{"lang":"zh-CN","type":null}}] should work 1`] = `
Array [
  "http://example.com/for-good?lang=zh-CN&type=&category=",
  Object {
    "body": "{\\"for\\":\\"the throne\\"}",
    "headers": Headers {
      "map": Object {
        "content-type": "application/json",
      },
    },
    "method": "POST",
  },
  undefined,
]
`;

exports[`http PUT http://example.com with [] should work 1`] = `
Array [
  "http://example.com",
  Object {
    "method": "PUT",
  },
  undefined,
]
`;
